pkg/streamingpromql/types/fpoint_ring_buffer.go

```diff
@@ -237,6 +239,49 @@ func (b *FPointRingBuffer) AnyAtOrBefore(maxT int64) bool {
 	return b.points[b.firstIndex].T <= maxT
 }
 
+// ChangesAtOrBefore returns the number of changes between subsequent points in this ring buffer from the first
+// point to the point with timestamp less than or equal to maxT.
+func (b *FPointRingBuffer) ChangesAtOrBefore(maxT int64) (float64, bool) {
+	fHead, fTail := b.UnsafePoints(maxT)
+
+	haveFloats := len(fHead) > 0 || len(fTail) > 0
+
+	if !haveFloats {
+		// Prometheus' engine doesn't support histogram for `changes` function yet,
+		// therefore we won't add that yet too.
+		return 0, false
+	}
+
+	changes := 0.0
+	// The first point in the buffer.
+	prev := b.First().F
+
+	// Comparing the point with the point before it.
+	accumulate := func(points []promql.FPoint) {
+		for _, sample := range points {
+			current := sample.F
+			if current != prev && !(math.IsNaN(current) && math.IsNaN(prev)) {
+				changes++
+			}
+			prev = current
+		}
+	}
+
+	// The points buffer is wrapped around, therefore we need to check changes starting from the buffer's head
+	// and then continue to the tail.
+	if len(fHead) > 0 && len(fTail) > 0 {
+		// We don't need to compare the first point with itself, hence start from fHead[1:]
+		accumulate(fHead[1:])
+		accumulate(fTail)
+	}
+
+	if len(fHead) > 0 && len(fTail) == 0 {
+		// We don't need to compare the first point with itself, hence start from fHead[1:]
+		accumulate(fHead[1:])
+	}
```

Comment:
This is consistent with how we use accumulate elsewhere. I don't think there is much overhead in range when the len is 0 vs using an if.
```suggestion
	// The points buffer is wrapped around, therefore we need to check changes starting from the buffer's head
	// and then continue to the tail.
	// We don't need to compare the first point with itself, hence start from fHead[1:]
	accumulate(fHead[1:])
	accumulate(fTail)
```

Reply:
You are right. We can simplify that. Updated in a7d04746f58fc416884ba3fbeb1f3536418eabfb